avatar: Show placeholder on image load error

vinodsingh-dev-gsoc opened the pull request:
Fixes #1558.
This change improves the user experience by displaying a consistent placeholder when an avatar image fails to load from its URL.
Adds an `errorBuilder` to the `AvatarImage` widget and includes a test case to verify the new behavior.
Co-authored-by: Chris Bobbe [cbobbe@zulip.com](mailto:cbobbe@zulip.com)

chrisbobbe commented:
Thanks for the PR! Please post before/after screenshots of the change. Please also change the commit sequence so the tests are added in the same commit that adds the behavior being tested.

vinodsingh-dev-gsoc commented:
| status| CombinedScreen | DM List | Profile Screen |
|--------|--------|--------|--------|
|Before| <img width="406" height="875" alt="before_fix_for_combined_list" src="https://github.com/user-attachments/assets/a2c9b4d7-c589-40ea-a192-1ada775ba012" />| <img width="384" height="864" alt="before_fix_for_dm_list" src="https://github.com/user-attachments/assets/903b1e9b-5938-4979-931e-097c0621a3b3" />|<img width="373" height="786" alt="before_fix_for_profile_page" src="https://github.com/user-attachments/assets/d5c53fca-5830-4b8e-a7af-c9afbdfe1acb" />|
| After fix | <img width="353" height="782" alt="after_fix_for_combined_list" src="https://github.com/user-attachments/assets/b6eb46d5-4749-4da3-af53-7b24f83b242b" /> | <img width="372" height="775" alt="after_fix_for_dm_list" src="https://github.com/user-attachments/assets/b46f6bd1-a96e-41cd-9158-9e98fc776339" />|<img width="365" height="785" alt="after_fix_profile_page" src="https://github.com/user-attachments/assets/0d59e3f2-8fc4-4bd0-804a-f126d067a4f0" />|

--------
As you can see in the "After" screenshot, the fix ensures that a placeholder is now displayed whenever an avatar URL is null or fails to load, instead of showing a blank space. This provides a clearer and more consistent UI for the user.

chrisbobbe reviewed (commented):
Thanks! Small comments below.

chrisbobbe commented:
I see you've pushed some changes, but I'm not sure if you intend them to be ready for review; I assume you'll comment here when ready. 🙂

vinodsingh-dev-gsoc commented:
| Before (Bot avatar URL fails)| After Fix (Shows Placeholder) |
|--------|--------|
| <img width="382" height="849" alt="before_fix_for_combined_list_bot_logo_not_loaded_light_mode" src="https://github.com/user-attachments/assets/13ac1384-a838-4297-9f88-150dd7223e4d" />| <img width="418" height="880" alt="after_fix_for_combined_list_bot_logo_not_loaded" src="https://github.com/user-attachments/assets/8ec6eee2-76db-40d9-ba28-e52c741b3e79" /> |
| <img width="399" height="853" alt="before_fix_for_combined_list_bot_logo_not_loaded_dark_mode" src="https://github.com/user-attachments/assets/f76a85f5-f37a-4f33-8aa6-b7f5b5364e16" />| <img width="418" height="907" alt="after_fix_for_combined_list_bot_logo_not_loaded_DarkMode" src="https://github.com/user-attachments/assets/60dc390b-aec8-4707-93a1-837f0c34de2a" />|

vinodsingh-dev-gsoc commented:
Hi @alya and @chrisbobbe,

Thanks for the detailed feedback! I've now updated the PR description with the new screenshots to better demonstrate the visual changes.

As requested, I've included before/after images for both light and dark themes, which show both regular avatars and the new placeholders in the same view.

I believe this addresses all the points. Eager for your review when you have a moment. Thank you!

chrisbobbe commented on lib/widgets/user.dart:
```diff
@@ -1,5 +1,4 @@
 import 'package:flutter/material.dart';
-
```
Please don't delete this blank line; it separates third-party imports from our own imports. (Here and elsewhere in the PR.)

chrisbobbe commented on lib/widgets/user.dart:
```diff
@@ -90,6 +89,7 @@ class AvatarImage extends StatelessWidget {
       avatarUrl.get(physicalSize),
       filterQuality: FilterQuality.medium,
       fit: BoxFit.cover,
+      errorBuilder: (_, _, _) => _AvatarPlaceholder(size: size),
```
and add a test for that.

chrisbobbe replied:
Same for the `user == null` case above that, for this part of the issue description:

> Or, pre-https://github.com/zulip/zulip-flutter/issues/1556, for a message from an unknown sender

chrisbobbe commented on test/widgets/user_test.dart:
```diff
@@ -78,5 +79,26 @@ void main() {
+        findsOneWidget,
+      );
+  debugNetworkImageHttpClientProvider = null;
+    });
```
nits:

- Empty line before a test
- Try to adjust indentation and spacing to follow the pattern of existing tests
- Use `check` instead of `expect`; see existing tests for examples

chrisbobbe reviewed (commented):
Thanks! Comments below, and please also address Greg's feedback about the commit sequence.

vinodsingh-dev-gsoc commented:
Hi @chrisbobbe, thanks for the feedback!

I've fixed the test failures (related to new_dm_sheet_test.dart) and also resolved the merge conflicts .
I've run tools/check locally, and all tests are passing now. This should be ready for another review. Thanks!

vinodsingh-dev-gsoc commented:
Hello @chrisbobbe  I've pushed the new updates addressing your feedbacks .
Could you please take another look when you have a moment? Thanks!

chrisbobbe commented on test/widgets/user_test.dart:
```diff
@@ -28,9 +27,12 @@ void main() {
+            accountId: eg.selfAccount.id,
+            child: AvatarImage(userId: user.userId, size: size ?? 30),
+          )
+      );
```
nit: keep indentation style (here and many other places):

```suggestion
      await tester.pumpWidget(
        TestZulipApp(accountId: eg.selfAccount.id,
          child: AvatarImage(userId: user.userId, size: size ?? 30)));
```

chrisbobbe reviewed (commented):
Thanks! Comments below.

Also some commit-message nits:

- In the first commit, `refactor(test):` doesn't follow our style for the prefix of the commit-message summary line. It should refer to a subsystem; see the doc: https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html#commit-summary-part-1
  - In this case I would choose `new-dm test:` or similar.
- In the second commit, check the style of the "Fixes" line: it should say `Fixes #1558.`

vinodsingh-dev-gsoc commented:
Hi @chrisbobbe!

I've just pushed the final updates, addressing all the nits and feedback (indentation, test helpers, and commit messages).

This should be ready for a final review. Thanks for all your guidance! 🙏

chrisbobbe commented on test/widgets/user_test.dart:
```diff
@@ -78,5 +83,51 @@ void main() {
+
+      await tester.pumpWidget(
+         TestZulipApp(accountId: eg.selfAccount.id,
+           child: AvatarImage(userId: badUser.userId, size: 30)));
```
Part of https://github.com/zulip/zulip-flutter/pull/1882#discussion_r2548238195 still isn't done; bump:

```suggestion
      await tester.pumpWidget(
        TestZulipApp(accountId: eg.selfAccount.id,
          child: AvatarImage(userId: badUser.userId, size: 30)));
```

gnprice commented on test/widgets/new_dm_sheet_test.dart:
```diff
@@ -254,17 +254,17 @@ void main() {
-      final icon = tester.widget<Icon>(find.descendant(
-        of: findUserTile(user),
-        matching: find.byType(Icon)));
+      Finder findInUserTile(User user, Finder finder) => find.descendant(of: findUserTile(user), matching: finder);
```
This is still wrongly indented. Chris already showed you how this should look; please reread more carefully the comment at https://github.com/zulip/zulip-flutter/pull/1882#discussion_r2548230190 .

gnprice replied:
nit: separately, this is too wide: necessary information appears after 80 columns.

gnprice commented on test/widgets/user_test.dart:
```diff
@@ -53,7 +58,7 @@ void main() {
 
-   testWidgets('absolute URL, larger size', (tester) async {
+    testWidgets('absolute URL, larger size', (tester) async {
       tester.view.devicePixelRatio = 2.5;
```
Please reread your commits before asking others to review them, and remove any changes that aren't doing something desired.

gnprice replied:
bump

gnprice commented on test/test_images.dart:
```diff
@@ -12,12 +13,18 @@ import 'package:flutter_test/flutter_test.dart';
+  } else {
+    httpClient.request.response
+      ..statusCode = HttpStatus.notFound
+      ..content = Uint8List(0);
```
(If we did want to add this feature to this function, then one thing we would definitely need to do at a minimum is to update its dartdoc — the documentation needs to be kept accurate.)

gnprice commented on test/widgets/user_test.dart:
```diff
@@ -78,5 +83,51 @@ void main() {
+      check(findPlaceholder).findsOne();
+    });
+
+    testWidgets('shows placeholder when user is not found', (WidgetTester tester) async {
```
nit: organize the tests in the same order as the code they're testing. This case comes before the other error cases in the implementation, so its test goes before the other error tests.

gnprice commented on test/widgets/user_test.dart:
```diff
@@ -78,5 +83,51 @@ void main() {
+      addTearDown(testBinding.reset);
+      await testBinding.globalStore.add(eg.selfAccount, eg.initialSnapshot());
+
+      const nonExistentUserId = 9999999;
```
Does this user ID really not exist? Will it for sure never exist in this test?

Seems like it *probably* won't, but the reader of the test shouldn't have to wonder. You can insert a check verifying that it doesn't exist, to let the reader be sure.

gnprice commented on test/widgets/user_test.dart:
```diff
@@ -78,5 +83,51 @@ void main() {
+      await tester.pumpWidget(
+        TestZulipApp(accountId: eg.selfAccount.id,
+          child: AvatarImage(userId: nonExistentUserId, size: 30)));
+      await tester.pumpAndSettle();
```
Avoid `pumpAndSettle`. See https://github.com/flutter/flutter/blob/master/docs/contributing/Style-guide-for-Flutter-repo.md#avoid-using-pumpandsettle — or the API doc for `pumpAndSettle` itself.

gnprice commented on test/widgets/user_test.dart:
```diff
@@ -78,5 +83,51 @@ void main() {
       debugNetworkImageHttpClientProvider = null;
     });
+
+    testWidgets('shows placeholder when image URL gives error', (WidgetTester tester) async {
```
nit: omit boring type
```suggestion
    testWidgets('shows placeholder when image URL gives error', (tester) async {
```

Note that the existing tests near this one all do the same.

gnprice reviewed (commented):
Thanks; comments below.

One commit-message nit:

    Co-authored-by: Chris Bobbe <cbobbe@zulip.com>
    Co-authored-by:  Greg Price <greg@zulip.com>

Cut extra space after colon.

Separately: why am I listed as a co-author? I don't think I wrote any of these changes 🙂

gnprice reviewed (commented):
Thanks! Just small comments below.
